### PR TITLE
Add support for cancelling individual UILocalNotification instances

### LINKED ISSUE #2920: [Request: PushNotificationIOS] Track and remove scheduled notifications
Is there a way to track and subsequently remove scheduled local notifications?

My naive usage wish would be something like:

```
var ref = PushNotificationIOS.scheduleLocalNotification({
      fireDate: new Date().toISOString(),
      alertBody: 'Something'
});

PushNotificationIOS.removeLocalNotification(ref);
```

### PR DESCRIPTION
**Pull request by andrewgleave**
Local notifications can be cancelled individually using an auto-generated identifier. Previously, the only way to cancel a notification was to cancel all notifications associated with the application using cancelAllLocalNotifications.

Also added a `loalNotification` and renamed existing `notification` and `registration` events to `remoteNotification`, `remoteRegistration` to make their intent clearer now both remote and local events are supported.

Fixes #2920, #2185

### CODE REVIEW

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 1]
```diff
@@ -117,6 +139,33 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
                                               body:notification.userInfo];
 }
 
+- (void)handleLocalNotificationReceived:(NSNotification *)notification
+{
+  UILocalNotification *local = [notification.userInfo objectForKey:@"notification"];
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: nit: use subscripting syntax where available, i.e. `notification.userInfo[@"notification"];`

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 2]
```diff
@@ -117,6 +139,33 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
...
+  NSArray *keys = [_scheduledLocalNotifications allKeysForObject:local];
+  NSString *identifier = [keys objectAtIndex:0];
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: `keys[0]` or `keys.firstObject`

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 3]
```diff
@@ -117,6 +139,33 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
...
+
+/**
+ * Remove all expired UILocalNotification instances that may still remain in _scheduledLocalNotifications
+ * since we can only explicity remove them if the app is active or the user taps the notification icon.
+ */
+- (void)handleApplicationDidBecomeActiveNotification:(__unused NSNotification *)notification
+{
+  for (NSString *identifier in [_scheduledLocalNotifications allKeys]) {
+    UILocalNotification *note = [_scheduledLocalNotifications objectForKey:identifier];
+    if([note.fireDate timeIntervalSinceReferenceDate] < [[NSDate date] timeIntervalSinceReferenceDate]) {
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: nit: space after if

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 4]
```diff
@@ -117,6 +139,33 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
...
+}
+
+/**
+ * Remove all expired UILocalNotification instances that may still remain in _scheduledLocalNotifications
+ * since we can only explicity remove them if the app is active or the user taps the notification icon.
+ */
+- (void)handleApplicationDidBecomeActiveNotification:(__unused NSNotification *)notification
+{
+  for (NSString *identifier in [_scheduledLocalNotifications allKeys]) {
+    UILocalNotification *note = [_scheduledLocalNotifications objectForKey:identifier];
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: `_scheduledLocalNotifications[identifier]`

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 5]
```diff
@@ -117,6 +139,33 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
...
+                                              body:userInfo];
+}
+
+/**
+ * Remove all expired UILocalNotification instances that may still remain in _scheduledLocalNotifications
+ * since we can only explicity remove them if the app is active or the user taps the notification icon.
+ */
+- (void)handleApplicationDidBecomeActiveNotification:(__unused NSNotification *)notification
+{
+  for (NSString *identifier in [_scheduledLocalNotifications allKeys]) {
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: nit: prefer dot syntax for properties, e.g. `_scheduledLocalNotifications.allKeys`, `note.fireDate.timeIntervalSinceReferenceDate`, etc.

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m (hunk 3 above)
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: This if statement is very hard to parse. I think it's saying "if note.fireDate is earlier than now" - is that correct? If so, I'd write it as:

    if (note.fireDate.timeIntervalSinceNow < 0) {

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 6]
```diff
@@ -212,13 +261,27 @@ - (NSDictionary *)constantsToExport
...
   [RCTSharedApplication() scheduleLocalNotification:notification];
+
+  NSString *identifier = [[NSUUID UUID] UUIDString];
+  [_scheduledLocalNotifications setValue:notification forKey:identifier];
+  callback(@[identifier]);
+}
+
+RCT_EXPORT_METHOD(cancelLocalNotification:(NSString *)identifier)
+{
+  UILocalNotification *notification = [_scheduledLocalNotifications objectForKey:identifier];
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: `_scheduledLocalNotifications[identifier];`

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 7]
```diff
@@ -212,13 +261,27 @@ - (NSDictionary *)constantsToExport
   [RCTSharedApplication() presentLocalNotificationNow:notification];
 }
 
-RCT_EXPORT_METHOD(scheduleLocalNotification:(UILocalNotification *)notification)
+RCT_EXPORT_METHOD(scheduleLocalNotification:(UILocalNotification *)notification callback:(RCTResponseSenderBlock)callback)
 {
   [RCTSharedApplication() scheduleLocalNotification:notification];
+
+  NSString *identifier = [[NSUUID UUID] UUIDString];
+  [_scheduledLocalNotifications setValue:notification forKey:identifier];
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: `_scheduledLocalNotifications[identifier] = notification;`

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 8]
```diff
@@ -212,13 +261,27 @@ - (NSDictionary *)constantsToExport
...
+  callback(@[identifier]);
```
Comment (nicklockwood) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: I think it's more common to use a monotonically increasing numeric value for identifiers like this in JS-world. Since these ids only seem to last for the lifetime of the app running, I don't see any reason why they need to be UUIDs, so an NSNumber would be cheaper.

**Comment by nicklockwood** [comment]
If I understand this code correctly it only allows you to cancel notifications set within the current run of the app, is that correct?

So if you set a local notification and then force-close the app (or the system closes it, or the phone is rebooted, etc) then there's no way to cancel that notification programmatically?

That seems like a bit of a design flaw.

**Comment by nicklockwood** [comment]
What about this:

When the module is initialized, enumerate through `RCTSharedApplication().scheduledLocalNotifications` and add each one to the `_scheduledLocalNotifications` dictionary, assigning a unique key to each.

Then, return that list of keys to the JS side using the `+(NSDictionary *)constantsToExport` method of the `RCTPushNotifications`. That way, the JS module will have a list of all registered notifications whenever it starts up, and will be able to unregister them retroactively.

It might also make sense to provide a way to get the data for a given notification (e.g. the message, or other info) so that the JS code can actually know what it's deleting, but perhaps that's out of scope for this PR.

**Comment by andrewgleave** [comment]
I amended the code to use your code convention. Is this convention specified anywhere, the contrib doc doesn't specify dot syntax or subscripting.

No idea how to get this pull request back to a usable state. Never having needed to do this, I've messed something up and seem to have made it worse. Do you want me to close and open a new PR?

**Comment by nicklockwood** [comment]
I'm not sure what you mean by "usable state" - it seems fine?

**Comment by andrewgleave** [comment]
I thought PRs had to be rebased to a single commit? I'm not sure how to get from its current state down to one.

**Comment by andrewgleave** [comment]
Re enumerating at startup: Even if we have a full list of the notifications registered, we don't know how they relate to an original identifier after termination. The newly-generated identifiers would no longer necessarily correspond to the one returned when the notification was created since some could have expired. Being able to access certain properties of the notifications could help to identify them but you'd need to write some code to reconcile/match them across the boundary.

I agree that loosing reference across the boundary after termination is a problem with the current approach, though. The only satisfactory solution really is for UILocalNotification to expose an identifier itself – but that's not likely to happen.

**Comment by nicklockwood** [comment]
No, there's no need to flatten your PRs - that's done automatically by our import process.

**Comment by nicklockwood** [comment]
OK, I'll pull this as-is. We can always add support for persistent notifications at a later date.

@facebook-github-bot import

**Comment by andrewgleave** [comment]
Ok cool. I'll submit any further improvements I make separately.

**Comment by satya164** [comment]
@andrewgleave Can you rebase this please?

**Comment by holmesal** [comment]
@andrewgleave this looks awesome! What's blocking this PR from being merged in?

**Comment by andrewgleave** [comment]
Only that I haven't got time to rebase this at the moment. Hopefully get around to it soon though.

**Comment by facebook-github-bot-2** [comment]
@andrewgleave do you have any updates for this pull request? It's been a while so we wanted to check in and see if you've looked at the requested changes.

**Comment by javache** [comment]
It looks like we've got an alternative approach now for cancelling local notifications so I'm going to close this PR. Let me know if there's anything else from this PR that you want to merge, but it might be better to open a new one for that.
